include/random_walks/shake_and_bake_walk.hpp

```
@@ -0,0 +1,169 @@
+
+    private:
+
+        Point get_direction(RandomNumberGenerator& rng)
+        {
+            VT z = GetDirection<Point>::apply(dim_, rng).getCoefficients();
+            MT I_cc = - A_row_k_ * A_row_k_.transpose();
+            I_cc.diagonal() += VT::Ones(dim_);
+            NT U = rng.sample_urdist();               
+            NT r = std::pow(U, NT(1)/NT(dim_-1)); 
+            NT cz = A_row_k_.dot(z);
+            VT z_tilde  = I_cc*z;
+            z_tilde *= r;
+            z_tilde /= std::sqrt(NT(1) - cz*cz);
+            
+            VT v = z_tilde - std::sqrt(NT(1) - r*r) * A_row_k_;
+            return Point(v);
+        }
+
+        void initialize(const Point& boundary_pt,
+                        int   facet_idx,
+                        RandomNumberGenerator& rng)
+        {
+            dim_ = P_.dimension();
+            m_ = P_.num_of_hyperplanes();
+            VT b=P_.get_vec();
+
+            NT kFacetEps = epsilon_;
+
+            // Input values 
+            p_ = boundary_pt;
+            if (facet_idx < 0) //if the facet not given we calculate 
+            {
+                facet_idx_ = -1;
+                for (int i = 0; i < m_; ++i) {
+                    VT ai = P_.get_facet_normal_vec(i);
+                    NT dist = std::abs(ai.dot(p_.getCoefficients()) - b.coeff(i));
+                    if (dist < kFacetEps) {
+                        facet_idx_ = i;
+                        break;
+                    }
+                }
+                if (facet_idx_ < 0)
+                    throw std::runtime_error("Boundary point not on any facet!");
+            }
+            else //we recheck if the facet and point compatible
+            {
+                VT ai = P_.get_facet_normal_vec(facet_idx);
+                NT dist = std::abs(ai.dot(p_.getCoefficients()) - b.coeff(facet_idx));
+                if (dist > kFacetEps)
+                    throw std::runtime_error("That is not facet index of the boundary point!");
```

How about the following: If it's not, then we could compute the index ourselves and if the point does not belong to any facet then we can throw a runtime error